fix: support for loading playground v2.5 single file checkpoint.

sayakpaul opened the pull request:
# What does this PR do?

Fixes: https://github.com/huggingface/diffusers/issues/7221.

Code to test:

```python
from diffusers import StableDiffusionXLPipeline, EDMDPMSolverMultistepScheduler
import torch

url = "https://huggingface.co/playgroundai/playground-v2.5-1024px-aesthetic/blob/main/playground-v2.5-1024px-aesthetic.safetensors"
pipeline = StableDiffusionXLPipeline.from_single_file(url)

pipeline.to(device="cuda", dtype=torch.float16)

prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
image  = pipeline(prompt=prompt, guidance_scale=3.0).images[0]
image.save("playground_test_image.png")
```

We need to be careful when loading the FP16 checkpoint. See internal thread: https://huggingface.slack.com/archives/C03HBN1C8CW/p1709703331962229. Currently, we parse it correctly from the single file checkpoint. However, when using the `diffusers` format, if someone specifies `variant="fp16"` it is going to incorrectly fetch the mean and std corresponding to the FP32 variant. During inference, it doesn't cause a huge problem, but it does when fine-tuning with the FP16 variant.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_7230). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

DN6 commented on src/diffusers/loaders/single_file_utils.py:
```diff
@@ -1381,65 +1402,86 @@ def create_scheduler_from_ldm(
+    is_playground_model=False,
 ):
-    scheduler_config = get_default_scheduler_config()
-    model_type = infer_model_type(original_config, model_type=model_type)
```
Can't we use `model_type` set to `Playground` instead of an `is_playground_model`?

DN6 commented on src/diffusers/loaders/single_file_utils.py:
```diff
@@ -1381,65 +1402,86 @@ def create_scheduler_from_ldm(
+        scheduler_config = get_default_scheduler_config()
+        model_type = infer_model_type(original_config, model_type=model_type)
 
-    global_step = checkpoint["global_step"] if "global_step" in checkpoint else None
```
In general I want to avoid trying to introduce a lot of nested conditional statements into single file again.

DN6 commented on src/diffusers/loaders/single_file_utils.py:
```diff
@@ -1223,14 +1229,29 @@ def create_diffusers_unet_model_from_ldm(
+    image_size=None,
+    scaling_factor=None,
+    torch_dtype=None,
+    is_playground_model=False,
```
I think we can use the `model_type` argument instead.

DN6 commented on src/diffusers/loaders/single_file.py:
```diff
@@ -49,6 +49,7 @@ def build_sub_model_components(
     model_type=None,
     image_size=None,
     torch_dtype=None,
+    is_playground_model=False,
```
I don't think we need this argument if we can infer it from the checkpoint keys.

nicksonredfield replied:
> I don't think we need this argument if we can infer it from the checkpoint keys.

is it possible to push this fix to A1111? really need it, tks!

sayakpaul commented:
@DN6 I have addressed your comments. The code snippet posted in the OP generates:

![playground_test_image](https://github.com/huggingface/diffusers/assets/22957388/042f9239-56c5-416c-81af-14efa472fb53)

DN6 commented on src/diffusers/loaders/single_file.py:
```diff
@@ -245,6 +252,7 @@ def from_single_file(cls, pretrained_model_link_or_path, **kwargs):
         load_safety_checker = (kwargs.pop("load_safety_checker", False)) or (
             passed_class_obj.get("safety_checker", None) is not None
         )
+        model_type = "Playground" if "edm_mean" in checkpoint and "edm_std" in checkpoint else model_type
```
Can we move inferring the model type into this function? Can update it to accept checkpoint as well. It would keep model type inference logic isolated to that function.  
https://github.com/huggingface/diffusers/blob/6d3e82c9cd00483104a6f405d5e5893c2bed5c20/src/diffusers/loaders/single_file_utils.py#L310

nicksonredfield replied:
> Can we move inferring the model type into this function? Can update it to accept checkpoint as well. It would keep model type inference logic isolated to that function.
> 
> https://github.com/huggingface/diffusers/blob/6d3e82c9cd00483104a6f405d5e5893c2bed5c20/src/diffusers/loaders/single_file_utils.py#L310

s it possible to push this fix to A1111? really need it, tks!

sayakpaul commented:
@DN6 done.

nicksonredfield commented:
> @DN6 done.

is it possible to push this fix to A1111? really need it, tks!

DN6 commented on src/diffusers/loaders/single_file_utils.py:
```diff
@@ -305,7 +307,11 @@ def is_valid_url(url):
 
-def infer_model_type(original_config, model_type=None):
+def infer_model_type(original_config, checkpoint=None, model_type=None):
+    if checkpoint is not None:
```
Since checkpoint is always passed I think we can move it to below the `model_type` check. This way if `model_type` gets set correctly if explicitly passed.

sayakpaul replied:
Should be done in 2be231cc.

DN6 commented on src/diffusers/loaders/single_file_utils.py:
```diff
@@ -305,7 +307,11 @@ def is_valid_url(url):
-def infer_model_type(original_config, model_type=None):
+def infer_model_type(original_config, checkpoint=None, model_type=None):
+    if checkpoint is not None:
+        if "edm_mean" in checkpoint and "edm_std":
```
```suggestion
        if "edm_mean" in checkpoint and "edm_std" in checkpoint:
```

DN6 commented on src/diffusers/loaders/single_file_utils.py:
```diff
@@ -1223,14 +1240,34 @@ def create_diffusers_unet_model_from_ldm(
+    )
 
-    vae_config = create_vae_diffusers_config(original_config, image_size=image_size, scaling_factor=scaling_factor)
+    if "edm_mean" in checkpoint and "edm_std" in checkpoint:
+        if checkpoint["edm_mean"] is not None and checkpoint["edm_std"] is not None:
+            edm_mean = checkpoint["edm_mean"].to(dtype=torch_dtype) if torch_dtype else checkpoint["edm_mean"]
+            edm_std = checkpoint["edm_std"].to(dtype=torch_dtype) if torch_dtype else checkpoint["edm_std"]
+            vae_config = create_vae_diffusers_config(
+                original_config,
+                image_size=image_size,
+                scaling_factor=scaling_factor,
+                latents_mean=edm_mean.flatten().tolist(),
+                latents_std=edm_std.flatten().tolist(),
+            )
+    else:
+        vae_config = create_vae_diffusers_config(original_config, image_size=image_size, scaling_factor=scaling_factor)
```
Could we do something like this here? 

```suggestion
    if model_type == "Playground":
        edm_mean = checkpoint["edm_mean"].to(dtype=torch_dtype) if torch_dtype else checkpoint["edm_mean"]
        edm_std = checkpoint["edm_std"].to(dtype=torch_dtype) if torch_dtype else checkpoint["edm_std"]

        edm_mean = edm_mean.flatten().tolist()
        edm_std = edm_std.flatten().tolist()

    else:
        edm_mean = None
        edm_std = None

    vae_config = create_vae_diffusers_config(
            original_config, 
            image_size=image_size, 
            scaling_factor=scaling_factor,
            latents_mean=edm_mean,
            latents_std=edm_std
    )
```

sayakpaul commented:
@DN6 up for another review :-)

nicksonredfield commented:
> @DN6 up for another review :-)

Recently I've been using the PG2 model for character and scene design for my novel, creating some great images, and I've found that the DPMA sampler and high amount of cue words are only available in the A1111 webui to realize the full potential of the PG model, and I was very eager to continue my creations with PG2.5 when it was released. But I found that comfyUI couldn't release its potential and the A1111 was buggy running the model.

I'm not a developer, so I can't understand the details of what you guys are discussing, and I don't know if you guys will push this fix to the A1111 branch, so if this model is able to run in A1111, I hope to let me know here or on HG's model release page.

My creations stem from your support and it is much appreciated.
![00257-3013793337](https://github.com/huggingface/diffusers/assets/66860084/0807e120-9f06-4334-abf9-80f3b2eb82e4)
![00086-263538876](https://github.com/huggingface/diffusers/assets/66860084/4187c06e-93a3-4660-b79f-7d31bd437abe)
![00310-229345664](https://github.com/huggingface/diffusers/assets/66860084/42c4f813-245c-44b0-80d3-f0844dab45d3)

sayakpaul commented:
> I'm not a developer, so I can't understand the details of what you guys are discussing, and I don't know if you guys will push this fix to the A1111 branch, so if this model is able to run in A1111, I hope to let me know here or on HG's model release page.

We won't be able to push the fix for now.

nicksonredfield commented:
> > I'm not a developer, so I can't understand the details of what you guys are discussing, and I don't know if you guys will push this fix to the A1111 branch, so if this model is able to run in A1111, I hope to let me know here or on HG's model release page.
> 
> We won't be able to push the fix for now.

tks for reply, I would be very thankful if it could be done later, looking forward that.

sayakpaul commented:
Gonna merge this after the CI is complete. @DN6 do you want me to add a slow test?
